src/Adapter/MSTest.TestAdapter/Execution/TestMethodRunner.cs

```diff
@@ -254,6 +249,46 @@ internal UnitTestResult[] RunTestMethod()
         return results.ToUnitTestResults();
     }
 
+    private bool TryExecuteITestDataSource(List<TestResult> results)
+    {
+        if (_test.DataType != DynamicDataType.ITestDataSource)
+        {
+            return false;
+        }
+
+        UTF.ITestDataSource? dataSource;
+        object?[]? data;
+        if (_test.SerializedData?.Length == 3
+            && Enum.TryParse(_test.SerializedData[0], out TestDataIdentifierStrategy _))
+        {
+            if (!int.TryParse(_test.SerializedData[1], out int dataSourceIndex)
+                || !int.TryParse(_test.SerializedData[2], out int dataIndex))
+            {
+                // TODO
+                throw new InvalidOperationException();
+            }
+
+            dataSource = _testMethodInfo.GetAttributes<Attribute>(false)?.OfType<UTF.ITestDataSource>().Skip(dataSourceIndex).FirstOrDefault();
+            if (dataSource is null)
+            {
+                // TODO
+                throw new InvalidOperationException();
```

Comment:
Shall I handle gracefully or throw unreachable?

Reply:
I think it shouldn't happen?